### PR TITLE
[Issue 15797] Add std.regex.splitterWithMatches

### PR DESCRIPTION
**Pull request by JackStouffer**
Add function to split a string with a regex pattern but keep the matches

### CODE REVIEW

File: std/regex/package.d [hunk 1]
```diff
@@ -1429,5 +1429,123 @@ public @trusted String[] split(String, RegEx)(String input, RegEx rx)
...
+        Range _input;
+        size_t _offset;
+        bool onMatch = false;
+        alias Rx = typeof(match(Range.init,RegEx.init));
+        Rx _match;
+
+        @trusted this(Range input, RegEx separator)
+        {//@@@BUG@@@ generated opAssign of RegexMatch is not @trusted
+            _input = input;
+            //separator.flags |= RegexOption.global;
```
Comment (wilzbach) [inline std/regex/package.d]: did you forget to remove this?

File: std/regex/package.d [hunk 2]
```diff
@@ -1429,5 +1429,123 @@ public @trusted String[] split(String, RegEx)(String input, RegEx rx)
...
+                //No more separators, work is done here
+                _offset = _input.length + 1;
+            }
+            else
+            {
+                if (!onMatch)
+                {
+                    //skip past the separator
+                    _offset = _match.pre.length;
+                    onMatch = true;
```
Comment (wilzbach) [inline std/regex/package.d]: onMatch = !onMatch ? (after the if)

File: std/regex/package.d [hunk 3]
```diff
@@ -1429,5 +1429,123 @@ public @trusted String[] split(String, RegEx)(String input, RegEx rx)
...
+    {
+    private:
+        Range _input;
+        size_t _offset;
+        bool onMatch = false;
+        alias Rx = typeof(match(Range.init,RegEx.init));
+        Rx _match;
+
+        @trusted this(Range input, RegEx separator)
+        {//@@@BUG@@@ generated opAssign of RegexMatch is not @trusted
```
Comment (wilzbach) [inline std/regex/package.d]: Did you report the bug?

**Comment by DmitryOlshansky** [comment]
Cool except for the name ;)

**Comment by JackStouffer** [comment]
What would be a better name in your opinion?

**Comment by DmitryOlshansky** [comment]
I'd probably just add an option keepSeparators to the original splitter

**Comment by JackStouffer** [comment]
That's probably better. I think a compile-time switch a-la `isPermutation` would be the best option

**Comment by DmitryOlshansky** [comment]
> I think a compile-time switch a-la isPermutation would be the best option

Yeah, that will do.

**Comment by JackStouffer** [comment]
@DmitryOlshansky Now using a compile time switch and updated the documentation.

Also, any particular reason why this only works with strings and not any character range with slicing?

**Comment by DmitryOlshansky** [comment]
> Also, any particular reason why this only works with strings and not any character range with slicing?

Not much it just need to be relaxed throughout std.regex. That being said there not much of ranges that are random access of char.

File: std/regex/package.d [hunk 4]
```diff
@@ -1420,6 +1459,22 @@ unittest
...
+{
+    import std.algorithm.comparison : equal;
+
+    auto pattern = regex(`([\.,])`);
+
+    assert("2003.04.05"
+        .splitter!(Yes.keepSeparators)(pattern)
+        .equal(["2003", ".", "04", ".", "05"]));
+
+    assert("10:00a.m."
```
Comment (DmitryOlshansky) [inline std/regex/package.d]: Would want a case where: a separator starts the string and ends the string.

**Comment by JackStouffer** [comment]
@DmitryOlshansky Added test case

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by JackStouffer** [comment]
Thanks!

**Comment by MartinNowak** [comment]
Flags should come with a `alias KeepSeparator = Flag!"keepSeparator"` shortcut, so that client code can use `KeepSeparator.yes/no` and doesn't need to import std.typecons.

**Comment by JackStouffer** [comment]
@MartinNowak Andrei is against such aliases, as you can use `Yes.keepSeparator` already with no aliases.

**Comment by MartinNowak** [comment]
You need to import std.typecons for that to work and will neither get auto-completion, nor documentation.
But well, I'll go with whatever you guys think makes sense. Annoying that it's not consistent across phobos, those flag parameters might deserve a small entry in https://dlang.org/dstyle.html.

**Pull request merged**